📖Add CAEP Extensible Template Processing for clusterctl (#3052)
kubernetes-sigs/cluster-api: docs/proposals/20200511-clusterctl-extensible-template-processing.md

  prefered templating mechanisms.

## Proposal

This section outlines the high level work that needs to be done in order to
accomplish the task of supporting other templating tools.

### User Stories

- As a clusterctl user, I would like to preserve the current behavior of using
  existing workload cluster templates with simple variable substitution from
  the clusterctl config and environment variables.
- As an infrastructure provider, I would like to provide templated yaml with
  defaults and have that be consumed by `clusterctl config cluster` to
  generate cluster manifests.
- As a developer, I would like to easily provide a way to extend clusterctl
  with my preferred templating solution.

### Implementation Details/Notes/Constraints

Currently the clusterctl command fetches yaml for the provider components
(yaml defining the management cluster components) and the workload cluster
templates.

#### UX for using Extensible Templating Tools
For the first iteration, we are opting to use flags to activate the use of a
specific templating tool as follows,

```
# Retrieve variables required by the preferred template processor

> clusterctl config cluster <cluster-name> --infrastructure <provider-name> --list-variables --experimental-template-engine=templateProcessor

Reviewer: As discussed in zoom, I don't think we need to add the `--experimental-template-engine=templateProcessor` option in the CLI in this phase since we won't be implementing any specific processors as part of this initial proposal.

Author: Thanks for the feedback @CecileRobertMichon! I've updated the CAEP to reflect that we are currently exposing library changes only so the existing behavior and UX won't change.

  * Source https://github.com/k14s/ytt
* Cue - Cue Data Constraint Language.
  * Source: https://github.com/cuelang/cue
* Dhall - Dhall Programming Configuration Language.
  * Source: https://github.com/dhall-lang/dhall-lang/blob/master/README.md
  * Golang Library: https://github.com/philandstuff/dhall-golang
  * Kuberentes Library: https://github.com/dhall-lang/dhall-kubernetes
* Helm Template
  * Doc: https://helm.sh/docs/helm/helm_template/
  * Code: https://github.com/helm/helm/blob/master/cmd/helm/template.go

## Summary

Clusterctl handles the lifecycle of Cluster API management and workload
clusters. It automates fetching the YAML files defining provider components
and the cluster templates used to define workload clusters. However, as more
providers contribute to CAPI and the variety of workload cluster templates
increase, it becomes harder for operators to manage all the cluster template
flavor permutations. Operators and providers are now leaning towards yaml
templating tools to properly manage the variety of cluster templates, and to
simplify the end user experience by limiting the amount of boilerplate YAML
that operators need to supply.

> 

Reviewer 2: ```suggestion
```

This proposal outlines the need to create an extensible templating mechanism
within clusterctl so that it can be used effectively with other templating
tools and continue to provide a seamless user experience.

## Motivation

> 

Reviewer 2: ```suggestion
```

To support a Cluster API that is cloud agnostic, operators need to deploy
workloads in a multi-cloud environment. Operators and providers may need to
create and _manage_ multiple cluster templates across a matrix of options and
configurations. To manage the explosion of yaml manifests, operators may want
to explore templating tools to minimize duplication and assert type
correctness.

> 

Reviewer 2: ```suggestion
```

Users looking to deploy workload clusters may be overwhelmed by the number of
variables needed to configure a workload cluster. Some of these options may be
circumvented via defaults provided by the operators as part of the templating
tool.

> 

Reviewer 2: ```suggestion
```

This proposal aims to extend clusterctl’s current templating mechanism of
simple variable substitution to support other templating formats natively,
thus benefiting the operators to deploy workload clusters with minimal
configuration.

### Goals

> 

Reviewer 2: ```suggestion
```

- Clusterctl will not be changing the way it gets values for yaml variables.

>   That is, it will get values from environment variables and the config file.

Reviewer 2: Does this mean that we want to use the existing variable-finding mechanics we have today and not modify them, or that we want to be backward compatible with the current templating?

Author: This means that we are not changing the existing variable-finding mechanics. That is, clusterctl will pull in variables from the environment and the config file only.

Reviewer 2: If we're not changing anything, should it be called out as a goal?

Author: I guess backward compatibility is the goal so as to ensure the existing method will still work so that no providers or other consumers need to make any changes.

- To provide an interface such that other templating solutions may be
  supported.
- To implement ytt as the first templating extension.

### Non-Goals

> - To replace existing mechanisms such as kustomize.

Reviewer 2: Apologies if this has been answered already, `kustomize` isn't available as a clusterctl templating mechanism today, or am I missing something?

> - To replace Clusterctl default mode (environment variables substitution

Reviewer 2: Can we give the default mode a name and add it to the glossary? It'd be great to reference it by a specific name in the whole proposal and it'd help when we talk about it as well.

cc @fabriziopandini

Reviewer 2: `Simple`?

  templating).

### Future Work
- To automatically detect template files for the appropriate template engine.
- To use the configuration file to determine which provider should use
  prefered templating mechanisms. For example,
 ```yaml
  yamlProcessors:
  - provider: aws
    processor: ytt
 ```

>   We will be leaning towards using a flag for the initial implementation.

Reviewer 2: This point is presenting a potential implementation solution, should we just remove the example, or remove it altogether?

## Proposal

This section outlines the high level work that needs to be done in order to
accomplish the task of supporting other templating tools.

### User Stories

> 

Reviewer 2: ```suggestion
```

> - As a clusterctl user, I would like to preserve the current behavior of using

Reviewer 2: Is preserving a behavior something that a user would know about?

  existing workload cluster templates with simple variable substitution from
  the clusterctl config and environment variables.
- As an infrastructure provider, I would like to provide templated yaml (e.g.
  ytt templates) with defaults and have that be consumed by
  `clusterctl config cluster` to generate cluster manifests.
- As an infrastructure provider, I would like to provide a set of ytt
  templates in a compressed file format such as zip, gzip or tar containing
  all the ytt template artifacts and have that be consumed by `clusterctl

>   config cluster` to generate cluster manifests.

Reviewer 2: We should probably only pick one compression mechanism to start with

Reviewer 2: Kubernetes releases just use `tar.gz`, I'd use the same

- As a developer, I would like to easily provide a way to extend clusterctl
  with my preferred templating solution.

> 

Reviewer 2: ```suggestion
```

### Implementation Details/Notes/Constraints

> 

Reviewer 2: ```suggestion
```

Currently the clusterctl command fetches yaml for the provider components
(yaml defining the management cluster components) and the workload cluster
templates.

#### UX for using Extensible Templating Tools

> We can use flags to activate the use of a specific templating tool as follows,

Reviewer 2: ```suggestion
For the first iteration, we're opting to use flags to activate the use of a specific templating tool as follows,
```

```
# Retrieve variables required by the ytt templates

> clusterctl config cluster <cluster-name> --infrastructure <provider-name> --list-variables --experimental-template-engine=ytt

Reviewer 2: Probably off-topic, `list-variables` feels like it should be a sub-command

Reviewer: it does, I was thinking the same thing above `clusterctl init --list-images` yesterday... and they both require credentials to be set which seems wrong since nothing is actually getting deployed (but yeah definitely off topic)

Author: off topic but @CecileRobertMichon now you don't have to require credentials for that command. https://github.com/kubernetes-sigs/cluster-api/pull/3059

3. `Process` is responsible for parsing the artifact blob of bytes and will
   return the final manifest with values retrieved from the `VariablesGetter`.

The flag (`--experimental-template-engine`) value can be read in when
instantiating the `TemplateClient` and the appropriate type can be created.
This follows a similar pattern to the `RepositoryClient` and `Repository`
types.

#### Notes
- Access to using the specific template engines will be available only as part
  of an experimental flag.

- The yaml processor libraries will be located in a `contrib` subfolder. For
  example, it could be located within `cluster-api/cmd/clusterctl`.

- Another method of defining a yaml processor instead of using its library
  would be to shell out to the templating engine binary/executable.
  The advantage of this is that the templating engine is not a dependency within
  the clusterctl codebase.
  The disadvantage is that the user needs to ensure the binary is installed
  separately which disrupts the UX of clusterctl. If each infrastructure
  provider provides different template styles, the user would need to install
  multiple templating tool binaries.

> ### Risks and Mitigations

Reviewer 2: ```suggestion

### Risks and Mitigations
```

- Some templating tools will require multiple files to be defined rather than
  a single yaml file. These artifacts will need to be “grouped” together to
  support current retrieval mechanisms. Currently, `clusterctl config cluster`
  retrieves templates from multiple sources such as ConfigMaps within a
  cluster, URL, Github Repository, Local Repository and even the overrides
  directory. To ensure compatibility, we’ll need to establish a compression
  format like tar, gzip or zip.

## Alternatives
- An alternative to allowing clusterctl handle templates would be to just use
  another tool upstream to generate the various templates with variables
  defined. Then use clusterctl to consume that template and use the default
  variable substitution templating solution.

## Upgrade Strategy

Any individual template engine libraries would be upgraded via go modules
preferably using semver standards.

## Additional Details

Currently, cluster-api-provider-azure is using kustomize to author their
cluster template flavors. See [issue 540] for more details. Maybe we can
collect input from them to understand their strategy if they need to

> manage their template variations.

Reviewer 2: Risk/Mitigation?

Author: I don't see this as much of a risk but rather another source of information. I can reach out to @CecileRobertMichon and have a quick discussion about their use case.

Reviewer: @wfernandes did you still want to sync about this before the proposal merges?

Author: @CecileRobertMichon Sure let's do a zoom call sometime today if you're available and we can sync up 🙂

## Implementation History

> - [wfernandes] 04/22/2020: Proposed idea in an issue or [community meeting]
> - [wfernandes/fabrizio.pandini] 04/27/2020: Compile a [CAEP Google Doc] following the CAEP template
> - [wfernandes] 05/06/2020: First round of feedback from community
> - [wfernandes] 05/11/2020: Open proposal PR
> - [wfernandes] 05/13/2020: Present proposal at a [community meeting]

Reviewer 2: ```suggestion
- 04/22/2020: Proposed idea in an issue or [community meeting]
- 04/27/2020: Compile a [CAEP Google Doc] following the CAEP template
- 05/06/2020: First round of feedback from community
- 05/11/2020: Open proposal PR
- 05/13/2020: Present proposal at a [community meeting]
```

Authors should go in the list above

## Open Questions

1. [wfernandes] We need to define a format or convention for retrieving multiple template files.
    Currently, clusterctl supports template retrieval from a ConfigMap in a
    cluster, URL, Github release artifacts and local file system. We default to
    retrieving provider cluster templates from github release assets. In order to
    avoid running into github quota limits on retrieving asset files we can pull a
    compressed file with all the template contents.

      As mentioned above in the *Constraints* and *Implementation* section a suggestion
      for convention would be
      `cluster-template-<template-engine-type>.<compression-format>`. Compression

>       format could be tar, gzip, zip, etc.

Reviewer 2: This section should be removed before merging the PR, add it to risk/mitigations

This follows a similar pattern to the `RepositoryClient` and `Repository`
types.

#### Notes
- Access to using the specific template engines will be available only as part
  of an experimental flag.

- The yaml processor libraries will be located in a `contrib` subfolder. For
  example, it could be located within `cluster-api/cmd/clusterctl`.

- Another method of defining a yaml processor instead of using its library
  would be to shell out to the templating engine binary/executable.
  The advantage of this is that the templating engine is not a dependency within
  the clusterctl codebase.
  The disadvantage is that the user needs to ensure the binary is installed
  separately which disrupts the UX of clusterctl. If each infrastructure
  provider provides different template styles, the user would need to install
  multiple templating tool binaries.
### Risks and Mitigations
- Support path for supporting extended templating tools such as ytt within
  clusterctl. That is, if there are templating issues, who owns them and how
  do we address them. This is for each templating engine that is included in
  clusterctl.  
  However since we’ve decided to put extended templating

>   libraries within a **contrib** folder, it will be the responsibility of the

Reviewer 3: In order to properly address this, IMO we should update our contributor's guide to rule contribs (this can be done in a separate PR).

ideally, contribs can be assimilated to experiments in terms of graduation/deprecation rules for ensuring that the new addition to the codebase is maintained at least up to its maturity (see https://cluster-api.sigs.k8s.io/contributing#experiments ).

Author: @fabriziopandini @vincepri Should we create a separate story for including guidelines for **contrib** folder in our `CONTRIBUTING.md`?

Author: I created this issue https://github.com/kubernetes-sigs/cluster-api/issues/3074 to track this work.

Reviewer 2: The `contrib` should really be the current `exp` folder. W.r.t to the text above, I don't think we've defined what's an experiment/contrib lifecycle for clusterctl.

  format like tar, gzip or zip.

## Alternatives
- An alternative to allowing clusterctl handle templates would be to just use
  another tool upstream to generate the various templates with variables
  defined. Then use clusterctl to consume that template and use the default
  variable substitution templating solution.

## Upgrade Strategy

Any individual template engine libraries would be upgraded via go modules
preferably using semver standards.

## Additional Details

Currently, cluster-api-provider-azure is using kustomize to author their
cluster template flavors. See [issue 540] for more details. Maybe we can
collect input from them to understand their strategy if they need to
manage their template variations.

## Implementation History

- [wfernandes] 04/22/2020: Proposed idea in an issue or [community meeting]
- [wfernandes/fabrizio.pandini] 04/27/2020: Compile a [CAEP Google Doc] following the CAEP template
- [wfernandes] 05/06/2020: First round of feedback from community
- [wfernandes ] 05/13/2020: Present proposal at a [community meeting]
- [wfernandes ] 05/11/2020: Open proposal PR

## Open Questions
<!-- Add your questions here -->

> 1. [wfernandes] How do we pass in values to the YamlProcessor? Currently,

Reviewer 4: This seems similar to the problem that say, [Puppet Hiera](https://puppet.com/docs/puppet/latest/hiera_intro.html), tried to solve with a pluggable model. Most frequent options I've seen are the desire to use GPG or a system like Hashicorp Vault for access to secrets. The current model of environment variables or values in clusterctl.yaml is a bit poor in that regards.

The AWS CLI recently added the ability to [shell out to an external program](https://puppet.com/docs/puppet/latest/hiera_intro.html) for the retrieval of values which is also a neat model.

Author: These are all good options that can be added in the future because one of the goals of this proposal is to not change how `clusterctl` reads in values.
After investigating for the specific use case of ytt, we can use the existing method of reading in values either from the `clusterctl` config file. Since `clusterctl` uses viper under the hood to read in the config file, it is able to read nested values as well. So for the example mentioned below, we can specify 
```
aws:
  secret_key: foo
  access_key: bar
``` 
in the `clusterctl` config file and clusterctl will read it in as `aws.secret_key`. So I believe this question can be resolved. 🙂

   
   For example,
   
   ```
   aws:
    secret_key: foo
    access_key: bar
   ```
   
   And ytt can reference these values as flags in the form `aws.secret_key`.
   Should we provide other means of variable input to clusterctl?
   
   Notes:
   Ytt is able to read environment variables. However, it does have some caveats.
   [More information can be found here][ytt-data-values].
   
2. [wfernandes] We need to define a format or convention for retrieving multiple template files.
    Currently, clusterctl supports template retrieval from a ConfigMap in a
    cluster, URL, Github release artifacts and local file system. We default to
    retrieving provider cluster templates from github release assets. In order to
    avoid running into github quota limits on retrieving asset files we can pull a
    compressed file with all the template contents.
    
      As mentioned above in the *Constraints* and *Implementation* section a suggestion
      for convention would be
      `cluster-template-<template-engine-type>.<compression-format>`. Compression
      format could be tar, gzip, zip, etc.
    
3. [wfernandes] Do we want to extend the usage of other yaml processors to

>    include infrastructure components? That is, do we want to stick with using

Reviewer 5: One concrete thing I'd like to see is a way to make certain portions of provider templates optional. For example, when bootstrapping a new management cluster, you typically need to use a `Secret` to contain the credentials you want to use when communicating with your infrastructure provider (vSphere, AWS, Azure, etc.). After pivoting the Cluster API infrastructure deployment to the real new management cluster, in many situations, you'll want to use credentials granted to the underlying infrastructure machine (e.g. an AWS IAM instance profile) instead of a static credentials `Secret`. We don't necessarily have to figure this out as part of this initial work, but it is something I'd like to get to eventually.

Reviewer 3: +1
Another possible use case is the management of controller flags
(see https://github.com/kubernetes-sigs/cluster-api/issues/3045)

Author: Currently, the configuration required by the infrastructure components are directly related to credential management. For example, `AWS_B64_ENCODED_CREDENTIALS`, `AZURE_CLIENT_SECRET_B64`, `VSPHERE_PASSWORD`. Since there seems to be another effort to address the sharing of secrets in the future and possibly along with the management of controller flags, I suggest as part of this proposal, we use the default yaml processor for the infrastructure components.
That is, `clusterctl init` will operate as usual with variable substitution. It will not take the flag `--experimental-template-engine`.
